app/controllers/application_controller.rb

```diff
@@ -331,28 +331,33 @@ def handle_theme
     resolve_safe_mode
     return if request.env[NO_CUSTOM]
 
-    theme_id = request[:preview_theme_id]&.to_i
+    theme_ids = []
+
+    if preview_theme_id = request[:preview_theme_id]&.to_i
+      theme_ids << preview_theme_id
+    end
 
     user_option = current_user&.user_option
 
-    unless theme_id
+    if theme_ids.blank?
       ids, seq = cookies[:theme_ids]&.split("|")
       ids = ids&.split(",")&.map(&:to_i)
-      if ids && ids.size > 0 && seq && seq.to_i == user_option&.theme_key_seq.to_i
-        theme_id = ids.first
+      if ids.present? && seq && seq.to_i == user_option&.theme_key_seq.to_i
+        theme_ids = ids if guardian.allow_themes?(ids)
       end
     end
 
-    theme_id ||= user_option&.theme_ids&.first
+    theme_ids = user_option&.theme_ids || [] if theme_ids.blank?
 
-    if theme_id && !guardian.allow_themes?(theme_id)
-      theme_id = nil
+    unless guardian.allow_themes?(theme_ids)
```

Comment:
It looks like we're checking this twice , once on line 346 and another time here.

Reply:
Yeah we need to do that twice. The first one to check the user can use the themes specified in the `theme_ids` cookie, and this is to check the user can use themes in the `user_options` table.